aiobotocore/tokens.py

```diff
@@ -0,0 +1,160 @@
+import asyncio
+import logging
+from datetime import timedelta
+
+import dateutil.parser
+from botocore.compat import total_seconds
+from botocore.exceptions import ClientError, TokenRetrievalError
+from botocore.tokens import (
+    DeferredRefreshableToken,
+    FrozenAuthToken,
+    SSOTokenProvider,
+    TokenProviderChain,
+    _utc_now,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def create_token_resolver(session):
+    providers = [
+        AioSSOTokenProvider(session),
+    ]
+    return TokenProviderChain(providers=providers)
+
+
+class AioDeferredRefreshableToken(DeferredRefreshableToken):
+    def __init__(
+        self, method, refresh_using, time_fetcher=_utc_now
+    ):  # noqa: E501, lgtm [py/missing-call-to-init]
+        self._time_fetcher = time_fetcher
+        self._refresh_using = refresh_using
+        self.method = method
+
+        # The frozen token is protected by this lock
+        self._refresh_lock = asyncio.Lock()
+        self._frozen_token = None
+        self._next_refresh = None
+
+    async def get_frozen_token(self):
+        await self._refresh()
+        return self._frozen_token
+
+    async def _refresh(self):
+        # If we don't need to refresh just return
+        refresh_type = self._should_refresh()
+        if not refresh_type:
+            return None
+
+        # Block for refresh if we're in the mandatory refresh window
+        block_for_refresh = refresh_type == "mandatory"
+        if block_for_refresh or not self._refresh_lock.locked():
+            async with self._refresh_lock:
+                await self._protected_refresh()
+
+    async def _protected_refresh(self):
+        # This should only be called after acquiring the refresh lock
+        # Another task may have already refreshed, double check refresh
+        refresh_type = self._should_refresh()
+        if not refresh_type:
+            return None
+
+        try:
+            now = self._time_fetcher()
+            self._next_refresh = now + timedelta(seconds=self._attempt_timeout)
+            self._frozen_token = await self._refresh_using()
+        except Exception:
+            logger.warning(
+                "Refreshing token failed during the %s refresh period.",
+                refresh_type,
+                exc_info=True,
+            )
+            if refresh_type == "mandatory":
+                # This refresh was mandatory, error must be propagated back
+                raise
+
+        if self._is_expired():
+            # Fresh credentials should never be expired
+            raise TokenRetrievalError(
+                provider=self.method,
+                error_msg="Token has expired and refresh failed",
+            )
+
+
+class AioSSOTokenProvider(SSOTokenProvider):
```

Comment:
I'd prefer this being in credentials.py like in botocore but no biggie can move later

Reply:
that will make later diffing easier

Reply:
nm deleting thread, i was confused